In [1]:
#Importing necessary Libraries
import pandas as pd

In [2]:
#Reading Data
df = pd.read_csv('vehicles_us.csv')
#Looking at Data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


12 Columns, 51525 Rows.
Mandatory columns seem to be (Price, Model, Condition, Fuel(economy?), Transmission, Type, Date Posted, Days Listed) 8/12
Columns with missing data are (Model Year, Cylinders, Odometer, Paint Color, Is 4WD) 4/12

Model Year will be changed to integer.
Fuel being an object will be further investigated.
Is 4WD should be a boolean.
Price and Cylinders will be investigated for better data type.
Date Posted will be converted to Date data type.

In [3]:
#Getting a sample of data
df.sample(20)

price  model_year                     model  condition  cylinders  \
26707  18990      2000.0             ford f-350 sd       good        8.0   
41623  17990         NaN        chevrolet traverse       good        6.0   
43882  16595      2012.0                    bmw x5  excellent        8.0   
44149  22500      2016.0       chevrolet silverado   like new        6.0   
13632  19950      2015.0                  ram 1500   like new        8.0   
20844   6995      2010.0             honda odyssey       good        6.0   
19594  15500         NaN  chevrolet silverado 1500  excellent        8.0   
15334   3200      2003.0              toyota camry       good        4.0   
37668   1500      2005.0   chrysler town & country       fair        4.0   
23223   7900      2015.0            subaru outback       good        4.0   
19395  14800      2014.0  chevrolet silverado 1500  excellent        6.0   
16047  15500      2011.0                ford f-150  excellent        8.0   
20913   4500      2004.0                ford f-150  excellent        8.0   
27690   6750      2012.0              chrysler 200  excellent        6.0   
33500   6997      2011.0            ford econoline       good       10.0   
38855  15998      2017.0          chevrolet malibu  excellent        4.0   
47092   1500      2000.0             honda odyssey       fair        6.0   
6279    9995      2008.0           chevrolet tahoe       good        8.0   
917     5995      2010.0            toyota corolla  excellent        4.0   
7633    2995      2002.0               ford taurus  excellent        6.0   

         fuel  odometer transmission      type paint_color  is_4wd  \
26707  diesel  118774.0    automatic     truck         NaN     1.0   
41623     gas   52688.0    automatic       SUV       white     1.0   
43882     gas   73705.0    automatic       SUV       white     1.0   
44149     gas   13000.0    automatic    pickup       white     NaN   
13632     gas   84882.0    automatic     truck       black     1.0   
20844     gas       NaN    automatic  mini-van       black     NaN   
19594     gas  150000.0    automatic    pickup        grey     1.0   
15334     gas       NaN    automatic     sedan       black     NaN   
37668     gas  196000.0    automatic  mini-van        blue     NaN   
23223     gas       NaN    automatic     wagon       green     1.0   
19395     gas       NaN    automatic     truck      silver     NaN   
16047     gas  156000.0    automatic     truck       white     1.0   
20913     gas  218000.0    automatic    pickup         NaN     1.0   
27690     gas   98135.0    automatic     sedan         red     NaN   
33500     gas  248135.0    automatic       bus       white     NaN   
38855     gas       NaN    automatic     sedan      custom     NaN   
47092     gas  230000.0    automatic  mini-van       green     NaN   
6279      gas  174897.0    automatic       SUV         NaN     1.0   
917       gas  144000.0    automatic     sedan         red     NaN   
7633      gas       NaN    automatic     sedan      silver     NaN   

      date_posted  days_listed  
26707  2018-09-20           48  
41623  2019-04-13           11  
43882  2019-01-27           11  
44149  2018-11-06            8  
13632  2018-10-27           21  
20844  2019-03-26           10  
19594  2019-02-10           28  
15334  2018-12-04           28  
37668  2018-07-10           12  
23223  2018-11-26          107  
19395  2018-06-07           57  
16047  2018-07-09           17  
20913  2018-11-15           10  
27690  2019-02-22           13  
33500  2019-01-26           29  
38855  2019-04-19           22  
47092  2019-02-16           31  
6279   2018-11-20            6  
917    2018-12-19            6  
7633   2018-05-15          107

In [4]:
#Printing missing values
print('Missing values:')
print(df.isna().sum())
print()

#Printing unique values in the 'is_4wd' column
print('Unique values in "is_4WD":')
print(df['is_4wd'].unique())

Missing values:
price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

Unique values in "is_4WD":
[ 1. nan]


### Fixing Data

In [5]:
#Changing 'date_posted' to datetime data type
df['date_posted'] = pd.to_datetime(df['date_posted'])

#Chagning 'model_year' to int data type
df['model_year'] = df['model_year'].fillna(-1).astype('int64')

#Filling missing values in 'is_4wd' and changing to int data type
df['is_4wd'] = df['is_4wd'].fillna(0).astype('int64')
df.sample(20)

price  model_year                       model  condition  cylinders  \
10867  17400        2012               toyota tundra  excellent        8.0   
43265  19500        1993                ford mustang  excellent        8.0   
12181   7550        2013               nissan maxima       good        6.0   
47959   4800          -1                 ford ranger  excellent        4.0   
22036   5500        2015               nissan sentra       good        4.0   
2343    2000        1997              toyota 4runner       good        6.0   
37915  23495        2014               toyota tacoma  excellent        6.0   
39500   2990        1997                 ford ranger  excellent        4.0   
12960   1800        2003                honda accord  excellent        4.0   
39749  14000        2006               toyota tacoma       good        6.0   
20229   4550        2004  jeep grand cherokee laredo       good        8.0   
41740   3000        2001               ford explorer       fair        8.0   
16347   5500        2008         jeep grand cherokee       good        8.0   
22336  15950        2014             gmc sierra 1500  excellent        8.0   
18711  17925        2016                 toyota rav4  excellent        4.0   
9680    6500        2006               dodge charger       good        6.0   
46573  13995        2004    chevrolet silverado 1500  excellent        8.0   
25139   3895        2006               nissan murano       good        6.0   
9954   14900        2010                  ford f-150   like new        8.0   
39632  10100        2012               dodge charger  excellent        6.0   

      fuel  odometer transmission    type paint_color  is_4wd date_posted  \
10867  gas   86551.0    automatic  pickup       white       1  2018-08-16   
43265  gas  116000.0       manual   coupe        blue       0  2018-06-04   
12181  gas  175000.0    automatic   sedan        grey       0  2018-12-20   
47959  gas       NaN    automatic   truck       white       0  2018-12-04   
22036  gas  116000.0    automatic   sedan       black       0  2019-03-03   
2343   gas   95000.0       manual     SUV         red       1  2019-04-03   
37915  gas  100802.0    automatic   truck        grey       1  2018-09-07   
39500  gas  204432.0    automatic   truck        blue       0  2018-05-03   
12960  gas  125000.0    automatic   sedan        grey       0  2018-08-30   
39749  gas  240000.0    automatic   truck      silver       1  2018-05-03   
20229  gas  199997.0    automatic     SUV        grey       1  2018-12-05   
41740  gas  164000.0    automatic     SUV       green       0  2018-05-12   
16347  gas  148578.0    automatic     SUV       black       1  2018-07-18   
22336  gas  135136.0    automatic   truck        blue       1  2019-02-07   
18711  gas   88353.0    automatic     SUV      silver       1  2018-05-11   
9680   gas  115204.0    automatic   sedan         red       0  2018-06-30   
46573  gas       NaN    automatic  pickup       black       1  2018-05-23   
25139  gas  214932.0    automatic   wagon      custom       0  2018-06-04   
9954   gas  121920.0    automatic   truck         NaN       1  2019-01-07   
39632  gas  108000.0    automatic   other      silver       0  2018-05-30   

       days_listed  
10867           15  
43265           56  
12181           61  
47959           47  
22036           10  
2343            37  
37915           32  
39500           44  
12960            4  
39749          177  
20229           48  
41740           47  
16347           25  
22336           49  
18711           61  
9680            35  
46573           30  
25139           48  
9954            42  
39632           35

### Enrich Data


In [6]:
df['manufacturer'] = df['model'].apply(lambda x: x.split()[0])
df.sample(20)

price  model_year                       model  condition  cylinders  \
35336  11900        2003               toyota tacoma  excellent        6.0   
41122  17490        2016              ford econoline       good        NaN   
25315  33990        2018                  ford f-150       good        8.0   
8158    7498        2013                  ford focus  excellent        4.0   
13490  14980        2003  chevrolet silverado 2500hd   like new        8.0   
21347   4999        2010                nissan rogue       good        4.0   
49088   2500        2006               nissan sentra       good        4.0   
51316   4999        2010             hyundai elantra       good        4.0   
13379   9800        2014                 ford fusion  excellent        4.0   
45210   3995        2011                nissan versa  excellent        4.0   
28815   3950        2005                   ford f150       good        8.0   
7892   15980        2008             chevrolet tahoe   like new        8.0   
1032    5500        1997               jeep wrangler       good        6.0   
33674   5200        2014                chrysler 200  excellent        4.0   
2806    9900        2007         chevrolet silverado       good        8.0   
16339   3400        2005     chrysler town & country  excellent        6.0   
36864   7995        2005                chrysler 300       good        8.0   
49995   2750        2002       chevrolet trailblazer  excellent        6.0   
34254  14977        2012               toyota tacoma  excellent        6.0   
8047    2895        1995            chevrolet camaro  excellent        6.0   

         fuel  odometer transmission       type paint_color  is_4wd  \
35336     gas  183000.0    automatic     pickup         NaN       1   
41122     gas  122974.0    automatic      truck         NaN       0   
25315  diesel   23603.0    automatic      truck      silver       1   
8158      gas   78128.0    automatic  hatchback        grey       0   
13490  diesel  212000.0    automatic      truck         NaN       1   
21347     gas  168000.0    automatic        SUV       black       0   
49088     gas  176000.0    automatic      sedan        grey       0   
51316     gas  106935.0    automatic      sedan       black       0   
13379     gas       NaN    automatic      sedan         NaN       0   
45210     gas  155000.0    automatic  hatchback         NaN       0   
28815     gas  217000.0    automatic      truck       black       1   
7892      gas  131878.0    automatic        SUV       black       1   
1032      gas       NaN    automatic        SUV         red       1   
33674     gas  109211.0    automatic      sedan       white       0   
2806      gas   97516.0    automatic      truck         NaN       1   
16339     gas  119722.0    automatic   mini-van      silver       0   
36864     gas  183080.0    automatic      sedan        grey       0   
49995     gas  215000.0    automatic        SUV        grey       1   
34254     gas  167430.0    automatic      truck         NaN       0   
8047      gas  161000.0    automatic      coupe       white       0   

      date_posted  days_listed manufacturer  
35336  2018-11-10           17       toyota  
41122  2018-06-19           24         ford  
25315  2018-06-11           22         ford  
8158   2018-09-30           35         ford  
13490  2019-01-06           73    chevrolet  
21347  2019-01-16           14       nissan  
49088  2018-11-24           70       nissan  
51316  2019-02-12           41      hyundai  
13379  2018-05-05           27         ford  
45210  2018-10-27           44       nissan  
28815  2018-06-13           11         ford  
7892   2018-05-31           11    chevrolet  
1032   2018-05-22           17         jeep  
33674  2018-10-22           74     chrysler  
2806   2019-02-14           66    chevrolet  
16339  2018-12-05            7     chrysler  
36864  2019-03-23           48     chrysler  
49995  2018-10-01           24    chevrolet  
34